n8n: 1.9.3 -> 1.46.0, repackage with pnpm.fetchDeps

gepbird opened the pull request:
Since https://github.com/NixOS/nixpkgs/pull/290715 was merged, it's much easier to package and maintain pnpm projects.

## Description of changes

- replace all the `node-env`, `node-composition` stuff with `pnpm.fetchDeps` and `pnpm.configHook`
- upgrade n8n to latest version
- move to pkgs/by-name
- set N8N_RELEASE_TYPE to stable
- remove old patches and build inputs that seem no longer necessary
  - `oracledb` and `@sap/hana-client` removal on aarch64: these don't exist in node_modules anymore and aarch64 build works
  - `pkgs.which`: probably was unused before, there were no references to it
  - `NIX_CFLAGS_COMPILE = "-DNODE_API_EXPERIMENTAL_NOGC_ENV_OPT_OUT";`: builds without this with node 20
- remove update script: there's no update script for similarly packaged apps, and updating is fairly simple
- format with nixfmt-rfc-style

## Issues I encountered and solved

### Error: cannot find '/build/source/node_modules/.pnpm/sqlite3@5.1.7/node_modules/sqlite3/build/Release/node_sqlite3.node

`pnpm.configHook` calls `pnpm install` with `--ignore-scripts` that skips downloading/building node-sqlite3 bindings. A solution is building it manually: (use `node-gyp` from nixpkgs rather than from pnpm installed node_modules to avoid other errors)

https://github.com/NixOS/nixpkgs/blob/077adff3243b078a0f4a49421038d6f3f16fdf37/pkgs/by-name/n8/n8n/package.nix#L65-L67

### turbo run build: Failed to create APIClient: Unable to set up TLS. builder error: No such file or directory

In turbo release 1.12.4, they switched to using `rustls-tls-native-roots` rust crate which tries to open a file from `SSL_CERT_FILE` environment variable, which is set to `/no-cert-file.crt` in a nix build environment. Adding `cacert` to build inputs will set this env var to a real file, fixing the issue.

### ERR_PNPM_BAD_PM_VERSION This project is configured to use v9.1.4 of pnpm. Your current pnpm is v9.1.1.

Patching `package.json` to replace pnpm version to the current fixes it:

https://github.com/NixOS/nixpkgs/blob/077adff3243b078a0f4a49421038d6f3f16fdf37/pkgs/by-name/n8/n8n/package.nix#L35

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gepbird committed:
n8n: 1.9.3 -> 1.46.0, repackage with pnpm.fetchDeps

gepbird committed:
n8n: add maintainer gepbird

gepbird commented on pkgs/by-name/n8/n8n/package.nix:
```diff
@@ -0,0 +1,109 @@
+  patchPhase = ''
+    runHook prePatch
+
+    jq '.packageManager = "pnpm@${pnpm.version}"' package.json | sponge package.json
```
This should be replaced with `removePackageManagerVersion` after https://github.com/NixOS/nixpkgs/pull/317739 is merged.

SuperSandro2000 replied:
That commit is no longer part of that PR.

SuperSandro2000 commented on pkgs/by-name/n8/n8n/package.nix:
```diff
@@ -0,0 +1,109 @@
+
+    mkdir -p $out/{lib,bin}
+    cp -r {packages,node_modules} $out/lib
+    ln -s $out/lib/packages/cli/bin/n8n $out/bin/n8n
+
+    wrapProgram $out/bin/n8n \
+      --set N8N_RELEASE_TYPE "stable"
```
```suggestion
    makeWrapper $out/lib/packages/cli/bin/n8n $out/bin/n8n \
      --set N8N_RELEASE_TYPE "stable"
```

SuperSandro2000 commented on pkgs/by-name/n8/n8n/package.nix:
```diff
@@ -0,0 +1,109 @@
+  };
+
+  # this can be removed when pnpm in nixpkgs is new enough for n8n
+  patchPhase = ''
+    runHook prePatch
+
+    jq '.packageManager = "pnpm@${pnpm.version}"' package.json | sponge package.json
+
+    runHook postPatch
+  '';
```
```suggestion
  postPatch = ''
    jq '.packageManager = "pnpm@${pnpm.version}"' package.json | sponge package.json
  '';
```

natsukium commented on pkgs/by-name/n8/n8n/package.nix:
```diff
@@ -0,0 +1,104 @@
+  dontStrip = true;
+
+  meta = with lib; {
+    description = "Free and source-available fair-code licensed workflow automation tool. Easily automate tasks across different services";
```
```suggestion
    description = "Free and source-available fair-code licensed workflow automation tool";
```
`meta.description` should be a single sentence.
https://nixos.org/manual/nixpkgs/unstable/#var-meta-description

gepbird replied:
Good catch, we could also use `meta.longDescription`.

natsukium commented:
Result of `nixpkgs-review pr 319310` at 32950b65 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>n8n</li>
  </ul>
</details>

natsukium reviewed (approved):
It works fine on x86_64-linux.
LGTM, thanks!

gepbird commented on pkgs/by-name/n8/n8n/package.nix:
```diff
@@ -0,0 +1,108 @@
+
+  # this can be removed when pnpm in nixpkgs is new enough for n8n
+  postPatch = ''
+    jq '.packageManager = "pnpm@${pnpm.version}"' package.json | sponge package.json
```
Nice! Also 1.46.0 is the latest, I'm updating.

gepbird commented:
In the last few hours I updated this package from 1.45.1 to 1.46.0 and removed the pnpm version patch.

Tested it on x86_64 and aarch64 linux, works as before.

natsukium reviewed (approved):
LGTM, thanks for your work!
